### DISCUSSION TITLE
Packaging Issues

### DISCUSSION THREAD

**Discussion by Vollstrecker** [pupnp/pupnp#386] [Ideas]
Hey,

I'm just working on replicating the functions of make dist and ran into some problems:

1. The cmake subfolder is not included in the tarball (my fault, just for completeness)
2. As bootstrap is run on make dist and the tarball includes only the configure script, cmake won't work at all with these as it reads the versions and defines for autoconfig.h from configure.ac.
3. autotools are not available every platform, so a tarball created there would differ from other.

I thought about some solutions and want to gather some opinions about them.

For problem 2:
1. Reading the values from configure.ac if present else read them configure. This would make the cmake/autoheader.cmake even harder to read and understand as it is now (yes, even small fixes take some time as I need to always search what happens where).
2. Creating a file that is read by both systems for the version. In this case the defines would still be read from one of the scripts, so no real benefit here.
3. Writing the versions directly into CMakeLists.txt. Then there would be two places to change them which sounds doable to me.
4. Creating an autoheader.in to not needing to read them from somewhere. This could be combined with solution 2 and 3.

My favorite would be combination of 3 and 4. If done so I have 2 more questions:
1. Should the version be written like in the configure.ac and calculated from that as it's done now, or directly as the resulting so-version?
2. Would it be better to create an autoheader.in for both systems and place them in one of the include-dirs? These values didn't change for a long time so I don't expect much work on them in the future.

Opinions?

**Comment by Vollstrecker** [comment]
I forgot problem 3:

1. we could just error out in this case to have always the same tarballs.
2. we could append -no-autotools or -cmake-only to the tarballs name

Option 2 would be my favorite as (no secret) I would love to see autotools nowhere except in the archives of history. I just don't know if the lib needs to be buildable somewhere where no cmake is available. And also in this case I would ask if the sources for the autotools stuff should be included instead of the generated ones, or this should be left out completely, if they should be included, there would be needed a different appendix I guess.

**Comment by mrjimenez** [comment]
> Hey,

Hi!

> I'm just working on replicating the functions of make dist and ran into some problems:

Good Thing!

> 1. The cmake subfolder is not included in the tarball (my fault, just for completeness)

Should be an easy fix.

> 2. As bootstrap is run on make dist and the tarball includes only the configure script, cmake won't work at all with these as it reads the versions and defines for autoconfig.h from configure.ac.

I think that the autotools tarball does include `configure.ac`. I agree, reading from `configure` is insane.

> 3. autotools are not available every platform, so a tarball created there would differ from other.

Certainly a problem. There would be no way to generate the `configure` script in such a platform.

> I thought about some solutions and want to gather some opinions about them.
> 
> For problem 2:
> 
> 1. Reading the values from configure.ac if present else read them configure. This would make the cmake/autoheader.cmake even harder to read and understand as it is now (yes, even small fixes take some time as I need to always search what happens where).
> 2. Creating a file that is read by both systems for the version. In this case the defines would still be read from one of the scripts, so no real benefit here.
> 3. Writing the versions directly into CMakeLists.txt. Then there would be two places to change them which sounds doable to me.
> 4. Creating an autoheader.in to not needing to read them from somewhere. This could be combined with solution 2 and 3.

Would that still be necessary if the tarballs include `configure.ac`? I think they do.

> My favorite would be combination of 3 and 4. If done so I have 2 more questions:
> 
> 1. Should the version be written like in the configure.ac and calculated from that as it's done now, or directly as the resulting so-version?

Do we have a link to the conclusion of that so-version discussion? As far as I understand, it is a mistake to mix versions and so-versions, but I must confess I don't remember our conclusion. Since we now got that right, I don't think it is necessary to change that.

> 2. Would it be better to create an autoheader.in for both systems and place them in one of the include-dirs? These values didn't change for a long time so I don't expect much work on them in the future.

What would you use to process the autoheader.in? Should be something that exists both in autotools and cmake systems.

> I forgot problem 3:
> 
> 1. we could just error out in this case to have always the same tarballs.

That would mean that tarballs could only be generated by autotools, and I understand that that is not what you want.

> 2. we could append -no-autotools or -cmake-only to the tarballs name
> 
> Option 2 would be my favorite as (no secret) I would love to see autotools nowhere except in the archives of history. I just don't know if the lib needs to be buildable somewhere where no cmake is available. And also in this case I would ask if the sources for the autotools stuff should be included instead of the generated ones, or this should be left out completely, if they should be included, there would be needed a different appendix I guess.

I am not up to date with relation to build systems wars, but I have no strong attachments to autotools. It just works and it is there from the beginning. But removing it will certainly introduce issues for some people. I have nothing against generating a `-cmake-only` tarball. In fact, it would benefit from both worlds and if autotools gets phased out, we would just stop generating it.

**Comment by Vollstrecker** [comment]
> Would that still be necessary if the tarballs include configure.ac? I think they do.

Jep, ark lists everything with first Uppercase then lowercase, it's there, which solves most problems.

> Do we have a link to the conclusion of that so-version discussion? As far as I understand, it is a mistake to mix versions and so-versions, > but I must confess I don't remember our conclusion. Since we now got that right, I don't think it is necessary to change that.

For sure it's a mistake, the question was "Do you want to write "AC_SUBST([LT_VERSION_IXML],       [12:2:1])" in configure.ac and "set (IXML_VERSION 12.2.1) in CMakeLists, or "set (IXML_VERSION 11.1.2)" as it would result after the math is done.

> What would you use to process the autoheader.in? Should be something that exists both in autotools and cmake systems.

Nope, more like upnpconfig.h with .h.in and .h.cm instead of generating them.

> I am not up to date with relation to build systems wars, but I have no strong attachments to autotools. It just works and it is there from the > beginning. But removing it will certainly introduce issues for some people. I have nothing against generating a -cmake-only tarball. In
> fact, it would benefit from both worlds and if autotools gets phased out, we would just stop generating it.

k, so when autotools then as always, when no autotools then -cmake-only but with the autotools-source and the bootstrap-script, so these qould at least work on the other plattforms.
